File: EinsteinRiddle.pl
right_of(X, Y) :- X =:= Y+1.
left_of(X, Y)  :- right_of(Y, X).

next_to(X, Y)  :- right_of(X, Y).
next_to(X, Y)  :- left_of(X, Y).

solution(Street) :-
   /* There are 5 houses in the street, all of different colours; 
      and their owners all have different nationalities, smoke different brands,
      have different pets, and have different drinks */
    Street = [
        house(1, Color1, Nationality1, Smoke1, Pet1, Drink1),
        house(2, Color2, Nationality2, Smoke2, Pet2, Drink2),
        house(3, Color3, Nationality3, Smoke3, Pet3, Drink3),
        house(4, Color4, Nationality4, Smoke4, Pet4, Drink4),
        house(5, Color5, Nationality5, Smoke5, Pet5, Drink5)],
    member(house(_,red    ,brit     ,_           ,_      ,_      ), Street), /* The Brit lives in the red house */
    member(house(_,_      ,swede    ,_           , dogs  ,_      ), Street), /* The Swede lives with dogs */
    member(house(_,_      ,dane     ,_           ,_      ,tea    ), Street), /* The Dane drinks tea */
    member(house(GreenHouse, green ,_ ,_ ,_ ,_ ), Street), /* There is a green house... */
    member(house(WhiteHouse, white ,_ ,_ ,_ ,_ ), Street), /* ...and a white house... */
    left_of(GreenHouse,WhiteHouse),                        /* ...and the green one is left to the white one. */
    member(house(_,green  ,_        ,_           ,_      ,coffee ), Street), /* The person in the green house drinks coffee */
    member(house(_,_      ,_        ,pall_mall   , birds ,_      ), Street), /* The person who smokes Pall Mall rears birds */
    member(house(_,yellow ,_        ,dunhill     ,_      ,_      ), Street), /* The person in the yellow house smokes Dunhill */
    member(house(3,_      ,_        ,_           ,_      ,milk   ), Street), /* The person in the middle house drinks milk */
    member(house(1,_      ,norwegian,_           ,_      ,_      ), Street), /* The Norwegian lives in the first house. */
    member(house(Blendie ,_ ,_ ,blends ,_   ,_ ), Street), /* The person that smokes Blends... */
    member(house(CatLover,_ ,_ ,_      ,cats,_ ), Street), /* ...and the person who loves cats... */
    next_to(Blendie, CatLover),                            /* ...live next to each other */
    member(house(Dunhillie ,_ ,_ ,dunhill,_     ,_ ), Street), /* The person that smokes Dunhills... */
    member(house(HorseLover,_ ,_ ,_      ,horses,_ ), Street), /* ...and the person that loves horses... */
    next_to(Dunhillie, HorseLover),                            /* ...live next to each other */
    member(house(_,_      ,_     ,blue_master,_      ,beer   ), Street), /* The person smoking Blue Master drinks beer */
    member(house(_,_      ,german   ,prince      ,_      ,_      ), Street), /* The German smokes Prince */
    member(house(BlueOwner,blue ,_ ,_ ,_ ,_ ), Street),      /* The person in the Blue House... */
    member(house(Norwegian,_ ,norwegian ,_ ,_ ,_ ), Street), /* and the Norwegian... */
    next_to(BlueOwner,Norwegian),                            /* ...live next to each other */                                                 
    member(house(Blendie,_,_,blends,_,_),Street),            /* The person that smokes Blends... */
    member(house(WaterDrinker,_,_,_,_,water),Street),        /* And the person that drinks water... */
    next_to(Blendie,WaterDrinker),                           /* ...live next to each other */         
    member(house(_,_      ,_        ,_           , fish  ,_      ), Street). /* Someone has a fish */

/* Resolve with 
 *     solution(Street)
 * or
 *     solution(Street), member(house(_,FishOwner,_,_,_,_),Street)
 * Demonstrate uniqueness with
 *     aggregate_all(count, (solution(Street)), Count).
 */

File: EinsteinRiddle.py
#!/usr/bin/env python
# Based on https://artificialcognition.github.io/who-owns-the-zebra

# Requires the python-constraint package
from constraint import *

positions =     [ 1, 2, 3, 4, 5 ]
colors =        ['red', 'green', 'white', 'yellow', 'blue']
nationalities = ['Brit', 'Swede', 'Dane', 'Norwegian', 'German']
cigars =        ['Pall Mall', 'Dunhill', 'Blends', 'Blue Master', 'Prince']
pets =          ['dogs', 'cats', 'birds', 'horses', 'fish']
drinks =        ['milk', 'water', 'tea', 'beer', 'coffee']

problem = Problem()

criteria = colors + nationalities + cigars + pets + drinks

# Try to do this 
problem.addVariables(criteria, positions)

problem.addConstraint(AllDifferentConstraint(), colors)
problem.addConstraint(AllDifferentConstraint(), nationalities)
problem.addConstraint(AllDifferentConstraint(), cigars)
problem.addConstraint(AllDifferentConstraint(), pets) # This one includes the fact that one of them is a fish
problem.addConstraint(AllDifferentConstraint(), drinks)


problem.addConstraint(lambda e, r: e == r, ["Brit",  "red"])
problem.addConstraint(lambda s, d: s == d, ["Swede", "dogs"])
problem.addConstraint(lambda d, t: d == t, ["Dane",  "tea"])
problem.addConstraint(lambda g, w: w-g == 1, ["green", "white"])
problem.addConstraint(lambda g, c: g == c, ["green", "coffee"])
problem.addConstraint(lambda p, b: p == b, ["Pall Mall", "birds"])
problem.addConstraint(lambda y, d: y == d, ["yellow",  "Dunhill"])
problem.addConstraint(InSetConstraint([3]), ["milk"])
problem.addConstraint(InSetConstraint([1]), ["Norwegian"])
problem.addConstraint(lambda b, c: abs(b-c)==1, ["Blends", "cats"])
problem.addConstraint(lambda d, h: abs(d-h)==1, ["Dunhill", "horses"])
problem.addConstraint(lambda bm, b: bm==b, ["Blue Master", "beer"])
problem.addConstraint(lambda g, p: g==p, ["German", "Prince"])
problem.addConstraint(lambda n, b: abs(n-b)==1, ["Norwegian", "blue"])


solution = problem.getSolutions() # There might be more than one
print(f"Solutions found: {len(solution)}")

for i in positions:
    for x in solution[0]: # We just choose the first solution
        if solution[0][x] == i:
            print(str(i), x)
